java/vector/src/main/codegen/templates/DenseUnionVector.java

```diff
@@ -0,0 +1,520 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import io.netty.buffer.ArrowBuf;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.ReferenceManager;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.BaseValueVector;
+import org.apache.arrow.vector.ValueVector;
+import org.apache.arrow.vector.complex.StructVector;
+import org.apache.arrow.vector.compare.VectorVisitor;
+import org.apache.arrow.vector.types.Types;
+import org.apache.arrow.vector.types.UnionMode;
+import org.apache.arrow.vector.compare.RangeEqualsVisitor;
+import org.apache.arrow.vector.types.pojo.ArrowType;
+import org.apache.arrow.vector.types.pojo.FieldType;
+import org.apache.arrow.vector.util.CallBack;
+import org.apache.arrow.vector.util.TransferPair;
+
+<@pp.dropOutputFile />
+<@pp.changeOutputFile name="/org/apache/arrow/vector/complex/DenseUnionVector.java" />
+
+
+<#include "/@includes/license.ftl" />
+
+        package org.apache.arrow.vector.complex;
+
+<#include "/@includes/vv_imports.ftl" />
+        import io.netty.buffer.ArrowBuf;
+        import java.util.ArrayList;
+        import java.util.Collections;
+        import java.util.Iterator;
+        import org.apache.arrow.vector.compare.VectorVisitor;
+        import org.apache.arrow.vector.complex.impl.ComplexCopier;
+        import org.apache.arrow.vector.util.CallBack;
+        import org.apache.arrow.vector.ipc.message.ArrowFieldNode;
+        import org.apache.arrow.memory.BaseAllocator;
+        import org.apache.arrow.vector.BaseValueVector;
+        import org.apache.arrow.vector.util.OversizedAllocationException;
+        import org.apache.arrow.util.Preconditions;
+
+        import static org.apache.arrow.vector.types.UnionMode.Dense;
+
+
+
+/*
+ * This class is generated using freemarker and the ${.template_name} template.
+ */
+@SuppressWarnings("unused")
+
+
+/**
+ * A vector which can hold values of different types. It does so by using a StructVector which contains a vector for each
+ * primitive type that is stored. StructVector is used in order to take advantage of its serialization/deserialization methods,
+ * as well as the addOrGet method.
+ *
+ * For performance reasons, DenseUnionVector stores a cached reference to each subtype vector, to avoid having to do the struct lookup
+ * each time the vector is accessed.
+ * Source code generated using FreeMarker template ${.template_name}
+ */
+public class DenseUnionVector extends UnionVector {
```

Comment:
I don't think this relationship should necessarily hold without a larger refactoring.

Reply:
Sure. The implementations are separated now.